pkg/agent/stats/collector.go

```diff
@@ -104,36 +104,66 @@ func (m *Collector) Run(stopCh <-chan struct{}) {
-	anpStatsMap := map[types.UID]*statsv1alpha1.TrafficStats{}
+	acnpStatsMap := map[types.UID]map[string]*statsv1alpha1.TrafficStats{}
+	anpStatsMap := map[types.UID]map[string]*statsv1alpha1.TrafficStats{}
+
 	for ofID, ruleStats := range ruleStatsMap {
-		policyRef := m.networkPolicyQuerier.GetNetworkPolicyByRuleFlowID(ofID)
-		if policyRef == nil {
+		rule := m.networkPolicyQuerier.GetRuleByFlowID(ofID)
+
+		if rule == nil {
 			// This should not happen because the rule flow ID to rule mapping is
 			// preserved for at least 5 seconds even after the rule deletion.
-			klog.Warningf("Cannot find NetworkPolicy that has ofID %v", ofID)
+			klog.Warningf("Cannot find NetworkPolicy Rule that has ofID %v", ofID)
 			continue
 		}
-		klog.V(4).Infof("Converting ofID %v to policy %s", ofID, policyRef.ToString())
+
+		klog.V(4).Infof("Converting ofID %v to policy %s", ofID, rule.PolicyRef.ToString())
 
 		var statsMap map[types.UID]*statsv1alpha1.TrafficStats
-		switch policyRef.Type {
+		var ruleStatsMap map[types.UID]map[string]*statsv1alpha1.TrafficStats
+		switch rule.PolicyRef.Type {
 		case cpv1beta.K8sNetworkPolicy:
 			statsMap = npStatsMap
+			policyStats, exists := statsMap[rule.PolicyRef.UID]
+			if !exists {
+				policyStats = new(statsv1alpha1.TrafficStats)
+				statsMap[rule.PolicyRef.UID] = policyStats
+			}
+			policyStats.Bytes += int64(ruleStats.Bytes)
+			policyStats.Sessions += int64(ruleStats.Sessions)
+			policyStats.Packets += int64(ruleStats.Packets)
 		case cpv1beta.AntreaClusterNetworkPolicy:
-			statsMap = acnpStatsMap
+			ruleStatsMap = acnpStatsMap
+			lastRuleStats, exists := ruleStatsMap[rule.PolicyRef.UID]
+			if !exists {
+				lastRuleStats = make(map[string]*statsv1alpha1.TrafficStats)
+				ruleStatsMap[rule.PolicyRef.UID] = lastRuleStats
+			}
+			trafficStats, trafficStatsExists := lastRuleStats[rule.Name]
+			if !trafficStatsExists {
+				trafficStats = new(statsv1alpha1.TrafficStats)
+				lastRuleStats[rule.Name] = trafficStats
+			}
+			trafficStats.Bytes += int64(ruleStats.Bytes)
+			trafficStats.Packets += int64(ruleStats.Packets)
+			trafficStats.Sessions += int64(ruleStats.Sessions)
 		case cpv1beta.AntreaNetworkPolicy:
-			statsMap = anpStatsMap
-		}
+			ruleStatsMap = anpStatsMap
+			lastRuleStats, exists := ruleStatsMap[rule.PolicyRef.UID]
+			if !exists {
+				lastRuleStats = make(map[string]*statsv1alpha1.TrafficStats)
+				ruleStatsMap[rule.PolicyRef.UID] = lastRuleStats
 
-		policyStats, exists := statsMap[policyRef.UID]
-		if !exists {
-			policyStats = new(statsv1alpha1.TrafficStats)
-			statsMap[policyRef.UID] = policyStats
+			}
+			trafficStats, trafficStatsExists := lastRuleStats[rule.Name]
+			if !trafficStatsExists {
+				trafficStats = new(statsv1alpha1.TrafficStats)
+				lastRuleStats[rule.Name] = trafficStats
+			}
+			trafficStats.Bytes += int64(ruleStats.Bytes)
+			trafficStats.Packets += int64(ruleStats.Packets)
+			trafficStats.Sessions += int64(ruleStats.Sessions)
```

Comment:
Can we reduce the code redundancy here?

Reply:
There are three similar traffic stats add up in `collector.go`, I refactors them in to external function.